src/ordered.jl

```diff
@@ -8,34 +8,23 @@ When applying the algorithm to an image, the threshold matrix is repeatedly tile
 to match the size of the image. It is then applied as a per-pixel threshold map.
 Optionally, this final threshold map can be inverted by selecting `invert_map=true`.
 """
-struct OrderedDither{T<:AbstractMatrix} <: AbstractBinaryDither
+struct OrderedDither{T<:AbstractMatrix} <: AbstractDither
     mat::T
 end
 
-function binarydither!(
-    alg::OrderedDither, out::GenericGrayImage, img::GenericGrayImage; invert_map=false
-)
+function binarydither(alg::OrderedDither, img::GenericGrayImage; invert_map=false)
     # eagerly promote to the same eltype to make for-loop faster
     FT = floattype(eltype(img))
     if invert_map
         mat = @. FT(1 - alg.mat)
     else
         mat = FT.(alg.mat)
     end
+    matsize = size(mat)
 
-    # TODO: add Threads.@threads to this for loop further improves the performances
-    #       but it has unidentified memory allocations
-    @inbounds for R in TileIterator(axes(img), size(mat))
-        mat_size = map(length, R)
-        if mat_size == size(mat)
-            # `mappedarray` creates a readonly wrapper with lazy evaluation with `srgb2linear`
-            # so that original `img` data is not modified.
-            out[R...] .= @views img[R...] .> mat
-        else # boundary condition
-            mat_inds = map(Base.OneTo, mat_size)
-            out_inds = map(getindex, R, mat_inds)
-            out[out_inds...] .= @views img[out_inds...] .> mat[mat_inds...]
-        end
+    out = Matrix{Int}(undef, size(img)...)
+    @inbounds for i in CartesianIndices(img)
+        out[i] = img[i] > mat[mod1.(Tuple(i), matsize)...] ? INDEX_WHITE : INDEX_BLACK
```

Comment:
Say this is a 512x512 image, looping like this requires you to compute `mod1(5, matsize[2])` for 512 times, which is definitely expensive.

How about pre-compute a lookup table, e.g.,

```julia
mat_pos_lookup = map(n->[mod1(i, n) for i in max(size(img))], matsize)
```

and then directly read the result from the table? This should be much faster and it also enables SIMD (IIUC):

```julia
@inbounds @simd for I in CartesianIndices(img)
    pos = map(Tuple(I), mat_pos_lookup) do (i, lookup)
        @inbounds lookup[i]
    end
    out[I] = ifelse(img[I] > mat[pos...], INDEX_WHITE, INDEX_BLACK)
end
```